### DISCUSSION TITLE
How to customize third party dependencies when package-ing?

### DISCUSSION THREAD

**Discussion by kinghuang** [explosion/spaCy#9386] [Help: Coding & Implementations] [answered]
The [spacy package](https://spacy.io/api/cli#package) command [generates a list of third party dependencies](https://github.com/explosion/spaCy/blob/f87ae3cb7dcfa9d955d66ecbfb03aae8c58bd42e/spacy/cli/package.py#L280) and fills them into the meta requirements. For each dependency, the acceptable versions is [coded as a minor version range](https://github.com/explosion/spaCy/blob/f87ae3cb7dcfa9d955d66ecbfb03aae8c58bd42e/spacy/cli/package.py#L230) like `some_package>=1.0.0,<1.1.0` for example.

Can I customize the version range, or leave it out completely? I'm finding this tight version range difficult to work with when I don't want to limit minor version upgrades (e.g., allow `>=1.0.0,<2.0.0`), or when I'm developing the next minor release of a dependency (e.g., `1.1.0dev0`) that hasn't been released, yet. The generated requirements are too restrictive.

**Comment by adrianeboyd** [comment, marked as answer]
You can customize the version ranges however you'd like, but you'd need to rerun `python setup.py sdist` or `python setup.py bdist_wheel` to update the final packages after you edit `meta.json`.

We were trying to automate more of this (especially including `spacy-transformers`), since we noticed it was an issue for users, but the defaults might not be what you need for your package. And there are also still cases where you might need to add dependencies by hand that aren't expressed directly through the registered functions (`sentencepiece` for some transformer models, for example).

**Reply by kinghuang** [reply]
Ok. I can edit `meta.json` after the fact. But, it'd be much more preferable to somehow pass the right information to the `spacy package` command. And, other dependencies as you've noted!

**Reply by adrianeboyd** [reply]
You can pass the information in an existing `meta.json` with `-m`, but double-check how everything is merged together to be sure it's doing what you expect. I think it ought to skip any requirements that are already listed, but I'm not entirely sure about how the rest of the meta will be merged off the top of my head.

**Reply by kinghuang** [reply]
Ah, I forgot about that option. I'll try it out for this. Thanks!

**Reply by kinghuang** [reply]
Just confirming that I'm able to override the generated requirements by passing in a `meta.json` file with just the requirements. The rest of the meta.json contents is generated by the [spacy package](https://spacy.io/api/cli#package) command as before.

`spacy package training/model-best packages --meta meta.json …`

**meta.json:**

```json
{
  "requirements": [
    "en-enverus-oi-lib"
  ]
}
```

I'm also able to add other modules that don't register spacy factories to the requirements list.
